File: iOS/SnapFresh/Model/FarmersMarket.m
//
//  FarmersMarket.m
//  SnapFresh
//
//  Created by Marco Abundo on 8/8/15.
//  Copyright © 2015 shrtlist.com. All rights reserved.
//

@import AddressBookUI;
#import "FarmersMarket.h"

@interface FarmersMarket () // Class extension
@property (nonatomic, readonly) NSString *address;
@property (nonatomic, readonly) NSString *products;
@property (nonatomic, readonly) NSString *schedule;
@end

@implementation FarmersMarket

#pragma mark - Designated initializer

- (instancetype)initWithDictionary:(NSDictionary *)dictionary
{
    NSDictionary *marketDetails = dictionary[@"marketdetails"];
    NSArray *addressItems = [marketDetails[@"Address"] componentsSeparatedByString:@", "];
    
    NSString *street = @"";
    NSString *city = @"";
    NSString *state = @"";
    NSString *zip = @"";
    
    if (addressItems.count == 2) {
        city = addressItems[0];
        state = addressItems[1];
    }
    else if (addressItems.count == 3)
    {
        id thirdItem = addressItems[2];
        
        if ([self hasLeadingNumberInString:thirdItem]) {
            // third item is ZIP code
            city = addressItems[0];
            street = addressItems[1];
            zip = addressItems[2];
        }
        else
        {
            street = addressItems[0];
            city = addressItems[1];
            state = addressItems[2];
        }
    }
    else if (addressItems.count > 3)
    {
        street = addressItems[0];
        city = addressItems[1];
        state = addressItems[2];
        zip = addressItems[3];
    }
    
    // Create the address dictionary
    NSDictionary *addressDictionary = @{(NSString *)kABPersonAddressStreetKey:street,
                                        (NSString *)kABPersonAddressCityKey:city,
                                        (NSString *)kABPersonAddressStateKey:state,
                                        (NSString *)kABPersonAddressZIPKey:zip};
    
    NSString *googleLink = marketDetails[@"GoogleLink"];
    NSURLComponents *urlComponents = [NSURLComponents componentsWithString:googleLink];
    NSURLQueryItem *queryItem = [urlComponents.queryItems firstObject];
    NSString *value = queryItem.value;
    NSArray *latLongArray = [value componentsSeparatedByString:@" "];
    NSNumber *lat = latLongArray[0];
    NSNumber *lon = latLongArray[1];
    
    // Create the coordinate
    CLLocationCoordinate2D coordinate = kCLLocationCoordinate2DInvalid;
    
    if (lat && lon)
    {
        coordinate = CLLocationCoordinate2DMake([lat doubleValue], [lon doubleValue]);
    }
    
    // Call our parent's initializer
    self = [super initWithCoordinate:coordinate addressDictionary:addressDictionary];
    
    if (self)
    {
        NSString *postalAddressString = ABCreateStringWithAddressDictionary(self.addressDictionary, NO);
        
        _address = [postalAddressString stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]];

        _products = marketDetails[@"Products"];

        _schedule = marketDetails[@"Schedule"];
    }
    
    return self;
}

- (BOOL)hasLeadingNumberInString:(NSString *)s
{
    if (s)
        return [s length] && isnumber([s characterAtIndex:0]);
    else
        return NO;
}

#pragma mark - Override MKAnnotation protocol accessors

- (NSString *)title
{
    return self.marketName;
}

- (NSString *)subtitle
{
    return self.address;
}

@end

File: iOS/SnapFresh/Model/FarmersMarket.swift
//
//  FarmersMarket.m
//  SnapFresh
//
//  Created by Marco Abundo on 8/8/15.
//  Copyright © 2015 shrtlist.com. All rights reserved.
//

import AddressBookUI
import CoreLocation
import MapKit

class FarmersMarket: MKPlacemark {

    // Variable stored properties
    var marketName: String = ""
    
    // MARK: Designated initializer
    init(dictionary: NSDictionary) {
        let marketDetails = dictionary["marketdetails"] as! NSDictionary
        let addressItems = marketDetails["Address"]!.componentsSeparatedByString(", ")
        
        var street = ""
        var city = ""
        var state = ""
        var zip = ""
        
        if (addressItems.count == 2) {
            city = addressItems[0];
            state = addressItems[1];
        }
        else if (addressItems.count == 3) {
            let thirdItem = addressItems[2] as String
            
            if NSString.hasLeadingNumberInString(thirdItem) {
                // third item is ZIP code
                city = addressItems[0];
                street = addressItems[1];
                zip = addressItems[2];
            }
            else {
                street = addressItems[0];
                city = addressItems[1];
                state = addressItems[2];
            }
        }
        else if (addressItems.count > 3) {
            street = addressItems[0];
            city = addressItems[1];
            state = addressItems[2];
            zip = addressItems[3];
        }
        
        // Create the address dictionary
        let addressDictionary: [String : AnyObject] = [String(kABPersonAddressStreetKey):street,
                                            String(kABPersonAddressCityKey):city,
                                            String(kABPersonAddressStateKey):state,
                                            String(kABPersonAddressZIPKey):zip]
        
        let googleLink = marketDetails["GoogleLink"] as! String
        let urlComponents = NSURLComponents(string: googleLink)
        let queryItem = urlComponents!.queryItems!.first
        let value = queryItem!.value
        let latLongArray = value!.componentsSeparatedByString(" ")
        let latString = latLongArray[0].stringByTrimmingCharactersInSet(NSCharacterSet.letterCharacterSet())
        let finalLatString = latString.stringByReplacingOccurrencesOfString(",", withString: "")
        let lonString = latLongArray[1].stringByTrimmingCharactersInSet(NSCharacterSet.letterCharacterSet())

        let lat = NSNumberFormatter().numberFromString(finalLatString)!.doubleValue
        let lon = NSNumberFormatter().numberFromString(lonString)!.doubleValue
        
        // Create the coordinate
        var coord = kCLLocationCoordinate2DInvalid
        
        if (lat != kCLLocationCoordinate2DInvalid.latitude && lon != kCLLocationCoordinate2DInvalid.longitude) {
            coord = CLLocationCoordinate2D(latitude: lat, longitude: lon)
        }
        
        // Call our parent's initializer
        super.init(coordinate: coord, addressDictionary: addressDictionary)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: Override MKAnnotation protocol accessors

    override var title: String? {
        get {
            return self.marketName;
        }
    }

    override var subtitle: String? {
        get {
            let postalAddressString = ABCreateStringWithAddressDictionary(self.addressDictionary!, false)
            
            return postalAddressString.stringByTrimmingCharactersInSet(NSCharacterSet.newlineCharacterSet())
        }
    }

}
